### PR TITLE
perf: single-buffer `concat_ws` kernel

### PR DESCRIPTION
**Pull request by davidlghellin**
Optimize the concat_ws kernel to a reused String buffer + StringBuilder with each argument downcast once, replacing the per-row Vec<String> + join. The coerce_types / Utf8View / list-variant parity from #1775 is preserved unchanged; this only rewrites the kernel internals. Also covers two previously-untested edge cases: concat_ws() now errors (WRONG_NUM_ARGS) and an empty array argument returns "". Verified 27/27 against Spark JVM 4.1.1 and Sail.

### CODE REVIEW

File: python/pysail/tests/spark/function/features/concat_ws.feature [hunk 1]
```diff
@@ -111,3 +111,155 @@ Feature: concat_ws function
...
+        | NULL   |
+        | x\|y   |
+
+    Scenario: concat_ws with all-scalar arguments broadcasts over rows
+      When query
+        """
+        SELECT concat_ws(',', 'a', 'b') AS result
+        FROM VALUES (1), (2), (3) AS t(x)
+        """
+      Then query result
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply python/pysail/tests/spark/function/features/concat_ws.feature:164]: The rows are all identical (`a,b`), so row order is not observable here. Switched the
assertion to `Then query result ordered` to make the intent explicit.

File: python/pysail/tests/spark/function/features/concat_ws.feature [hunk 2]
```diff
@@ -111,3 +111,155 @@ Feature: concat_ws function
...
+        | NULL   |
+        | NULL   |
+
+    Scenario: concat_ws skips per-row NULL over a column
+      When query
+        """
+        SELECT concat_ws(',', v, 'X') AS result
+        FROM VALUES ('a'), (CAST(NULL AS STRING)), ('c') AS t(v)
+        ORDER BY v
+        """
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply python/pysail/tests/spark/function/features/concat_ws.feature:136]: Good call — made the NULL ordering explicit with `ORDER BY v NULLS FIRST`. The expected
output is unchanged (the NULL row's `X` must sort first), since Spark's default for ASC
is already NULLS FIRST; this just documents the dependency so the test doesn't rely on an
implicit engine default. Fixed.

File: crates/sail-function/src/scalar/string/spark_concat_ws.rs [hunk 3]
```diff
@@ -92,200 +93,152 @@ impl ScalarUDFImpl for SparkConcatWs {
...
+        row_buf.clear();
+        let mut first = true;
+        for arg in &values {
+            arg.write_row(row, sep, &mut row_buf, &mut first)?;
+        }
+        builder.append_value(&row_buf);
     }
 
-    Ok(Arc::new(StringArray::from(results)))
+    Ok(Arc::new(builder.finish()))
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply crates/sail-function/src/scalar/string/spark_concat_ws.rs:141]: Not applicable here. `return_type` is hardcoded to `Utf8` (it never returns `LargeUtf8`),
so the `StringBuilder` (i32 offsets) exactly matches the declared output type — there's no
inconsistency. `LargeUtf8`/`Utf8View` are accepted only as *inputs* and read through the
input enum; the output is always `Utf8` on purpose, to match Spark's `StringType` for
`concat_ws`. Returning `LargeUtf8` would diverge from Spark rather than fix anything.

The i32-offset 2GB cap is a generic Arrow limitation shared by every `Utf8`-returning Spark
string function in Sail (batched execution bounds it in practice), not something specific to
this kernel. No change.

File: crates/sail-function/src/scalar/string/spark_concat_ws.rs [hunk 4]
```diff
@@ -92,200 +93,152 @@ impl ScalarUDFImpl for SparkConcatWs {
...
     let num_rows = args[0].len();
-    let separator_arr = &args[0];
 
-    // Extract separator strings (None for null separators)
-    let separators = get_string_values(separator_arr)?;
+    // A null-typed separator makes every row null. Return an N-null column (not a
+    // single scalar) so the shape lines up with the other arguments.
+    if *args[0].data_type() == DataType::Null {
+        return Ok(new_null_array(&DataType::Utf8, num_rows));
+    }
```
[the remark this answers is not in the document]
Comment (davidlghellin) [reply crates/sail-function/src/scalar/string/spark_concat_ws.rs:114]: Fixed

**Review by linhr — APPROVED** [review]
Nice improvements! Thanks!
